### DISCUSSION TITLE
Flashing Tinxy 16amp smart switch with Tasmota

### DISCUSSION THREAD

**Discussion by NsinghP** [arendst/Tasmota#18463] [Q&A] [answered]
I am going to flash Tinxy 16amp smart switch with Tasmota. I understand the physical connections to flash the device. However I am a bit confused about Tasmota Settings as this page is instructing to do: (https://templates.blakadder.com/tinxy_1_node_16a.html). 
![Screenshot (200)](https://user-images.githubusercontent.com/69470381/233553691-551b2461-3cf1-4827-b223-2daa845a1b38.png)
 How do I go about implementing the following into my custom bin file that I plan on generating through TasmoCompiler :

setoption15 0 (default seems to be 1 "ON")
setoption68 0 (seems to be default value)
Pwmfrequency 1000 (default seems to be 977)
Pwmrange 1023 (seems to be defualt value)
pwm1 512 (cant figure out default value, gives "command error" when i enter pwm1 in console)

Backlog rule on system#boot do pwm1 512 endon; Rule 1 1 <---Not sure but I am thinking of using manage file system feature of tasmota to create new file and enter this rule as is in there.

Question: Do I even need to integrate these into custom bin file or Should I make these changes after flashing the device through serial console? 

any pointers on integrating these options into custom file will be appreciated.

**Comment by barbudor** [comment, marked as answer]
> Question: Do I even need to integrate these into custom bin file or Should I make these changes after flashing the device through serial console?

You don't need.
Instructions there are intended as commands to be entered in the console after flashing.
There could be advantage to burn them into the binary but I suggest you first get experience with the device before going that way

As for the startup rule, it's lighter to leave it as a rule than using the file system and autoexec.bat.

**Reply by NsinghP** [reply]
> > Question: Do I even need to integrate these into custom bin file or Should I make these changes after flashing the device through serial console?
> 
> You don't need. Instructions there are intended as commands to be entered in the console after flashing. There could be advantage to burn them into the binary but I suggest you first get experience with the device before going that way
> 
> As for the startup rule, it's lighter to leave it as a rule than using the file system and autoexec.bat.

Thanks for your input. So it is my understanding that all I Ineed to do is enter following three lines, one after the other, in console: 

Setoption15 0
pwmfrequency 1000
Backlog rule on system#boot do pwm1 512 endon; Rule 1 1

and I will be done!

If i save the configuration file after making above changes, will they be saved and restored when I use this configuration file or will I have to manually reenter them again?

**Reply by NsinghP** [reply]
> > ...There could be advantage to burn them into the binary but I suggest you first get experience with the device before going that way
> 
> As for the startup rule, it's lighter to leave it as a rule than using the file system and autoexec.bat.

I am going to follow your advice and enter required customization using the console. Just out of curiosity I tried figuring out how to hard-code the above changes at time of compilation and came up with following: 

#define LIGHT_MODE   false              // [SetOption15] Switch between commands PWM or COLOR/DIMMER/CT/CHANNEL

#define LIGHT_CHANNEL_MODE     false             // [SetOption68] Enable multi-channels PWM instead of Color PWM

#define USER_BACKLOG "Rule1 1" on system#boot do pwm1 512 endon;<---EDIT: this line did not work and returned an error on tasmocompiler, but this line ---> #define USER_RULE1 "on system#boot do pwm1 512 endon" worked in the sense that it did not return an error and bin file was successfully compiled. (I will know if it works only after I have flashed the device with this file.)

Please Please correct any syntax/logic errors in above lines.

EDIT2: I think it worked here is the output of rule command in the console: 

16:53:39.089 RSL: RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":32,"Free":479,"Rules":"on system#boot do pwm1 512 endon"}}

Does this mean the rule is working? Did I do it right during compilation?


I still  have not figured out how to hardcode "Pwmfrequency 1000" and "Pwmrange 1023". I think following lines may have something to do with it, but it is just a guess:


//  #define USE_PCA9685                            // [I2cDriver1] Enable PCA9685 I2C HW PWM Driver - Must define I2C Address in #define USE_PCA9685_ADDR below - range 0x40 - 0x47 (+1k4 code)
//    #define USE_PCA9685_ADDR 0x40                // Enable PCA9685 I2C Address to use (Must be within range 0x40 through 0x47 - set according to your wired setup)
//    #define USE_PCA9685_FREQ 50                  // Define default PWM frequency in Hz to be used (must be within 24 to 1526) - If other value is used, it will rever to 50Hz

![Tinxy Tamota Flash](https://user-images.githubusercontent.com/69470381/233626277-6c73005a-a3e3-4752-8655-bbc0e6dc933d.jpeg)
![Screenshot (201)](https://user-images.githubusercontent.com/69470381/233626332-ca4bb975-e63f-4df8-a1ab-2f3f729ac0f1.png)

PS: The only manual command I had to enter was " pwmfrequency 1000 " and any pointers on how to hard code even this command will be appreciated.

EDIT: I was poking around in tasmota.h file and possibly found the hard-codded pwmfrequency values:

const uint32_t PWM_RANGE = 1023;            // 255..1023 needs to be devisible by 256
//const uint16_t PWM_FREQ = 1000;             // 100..1000 Hz led refresh
//const uint16_t PWM_FREQ = 910;              // 100..1000 Hz led refresh (iTead value)
const uint16_t PWM_FREQ = 977;              // 100..4000 Hz led refresh

is there a way to //the last line and un// the second line using user_config_override.h file method?

**Reply by barbudor** [reply]
Writing your personnal setting into Tasmota code is a bad idea. Everytime you will want to upgrade you will have to re-apply the changes again.
There are other solutions.

> #define USER_RULE1 "on system#boot do pwm1 512 endon" 

That is working for adding the rule but as you can see is the result, the rule is still off as you have to perform `Rule1 1` to enable it
You can also add in the binary a set of commands that would be executed on 1st boot:
```
#define USER_BACKLOG "Rule1 1;Pwmfrequency 1000;Pwmrange 1023"
```

There is one thing you've missed about burning settings into your binary is the template
If you want the device template to be in the binary you need to add those 2 lines:
```
#define MODULE USER_MODULE
#define TEMPLATE "{\"NAME\":\"Tnxy16A\",\"GPIO\":[32,0,0,0,160,224,0,0,288,0,416,0,0,0],\"FLAG\":0,\"BASE\":18}"
```
Note that as `"` is already the symbol for strings, when using it inside a string you need to escape it with a backslash so the compiler include the `"` character and doesn't interpret it as end/start of string

**Reply by NsinghP** [reply]
> Writing your personnal setting into Tasmota code is a bad idea. Everytime you will want to upgrade you will have to re-apply the changes again. There are other solutions.
> 
> > #define USER_RULE1 "on system#boot do pwm1 512 endon"
> 
> That is working for adding the rule but as you can see is the result, the rule is still off as you have to perform `Rule1 1` to enable it You can also add in the binary a set of commands that would be executed on 1st boot:
> 
> ```
> #define USER_BACKLOG "Rule1 1;Pwmfrequency 1000;Pwmrange 1023"
> ```
> 
> There is one thing you've missed about burning settings into your binary is the template If you want the device template to be in the binary you need to add those 2 lines:
> 
> ```
> #define MODULE USER_MODULE
> #define TEMPLATE "{\"NAME\":\"Tnxy16A\",\"GPIO\":[32,0,0,0,160,224,0,0,288,0,416,0,0,0],\"FLAG\":0,\"BASE\":18}"
> ```
> 
> Note that as `"` is already the symbol for strings, when using it inside a string you need to escape it with a backslash so the compiler include the `"` character and doesn't interpret it as end/start of string

Thanks, using above as cue, this is what I have come up with:

#define MODULE            USER_MODULE
#define USER_TEMPLATE "{\"NAME\":\"Tinxy16A-1\",\"GPIO\":[32,0,0,0,160,224,0,0,288,0,416,0,0,0],\"FLAG\":0,\"BASE\":18}"
#define LIGHT_MODE false 
#define LIGHT_CHANNEL_MODE false 
#define USER_BACKLOG "Rule1 1;Pwmfrequency 1000;Pwmrange 1023;on system#boot do pwm1 512 endon "
#define USE_EMULATION_HUE
#define EMULATION         EMUL_HUE
#define DEVICE_NAME       "Air Conditioner Switch"  
#define FRIENDLY_NAME     "First Air Conditioner Switch"

Please point out any errors before I flash the device. 

(Please ignore any errors (Specifically the Slash\) in #define User_Template line as those errors are either in cut and paste or due to html formatting in the forum, I still haven't figured it out why it does that every time i paste from notepad on windows11.)

**Reply by barbudor** [reply]
You didn't get the point about `\"`
Rule was at the right place in USER_RULE1. Why moving it to USER_BACKLOG and using a wrong syntax?

**Reply by NsinghP** [reply]
> You didn't get the point about `\"` Rule was at the right place in USER_RULE1. Why moving it to USER_BACKLOG and using a wrong syntax?

Sorry, mistake in my understanding, here is the next take on my understanding: (and I may be mistaken again :( but please bear with me):

```
#define MODULE            USER_MODULE
#define USER_TEMPLATE "{\"NAME\":\"Tinxy16A-1\",\"GPIO\":[32,0,0,0,160,224,0,0,288,0,416,0,0,0],\"FLAG\":0,\"BASE\":18}"
#define LIGHT_MODE false 
#define LIGHT_CHANNEL_MODE false
#define USER_RULE1 "on system#boot do pwm1 512 endon"
#define USER_BACKLOG "Rule1 1;Pwmfrequency 1000;Pwmrange 1023 "

```

Thanksssssssssss. above worked and I have proof: 

11:48:39.425 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":32,"Free":479,"Rules":"on system#boot do pwm1 512 endon"}}

NOTE: If you are powering the device from Mains then it is very difficult to do the OTA flash on the device if you don't have the rule  "on system#boot do pwm1 512 endon" activated, as the device keeps getting reset every 30 seconds and does not give enough time to enter wifi connection information. I had to power my already "Tasmota flashed" device using 3v3 and ground headers to be able to do OTA upgrade to latest " "on system#boot do pwm1 512 endon" rule enabled firmware.  (This problem may also be referenced in another discussion (https://github.com/arendst/Tasmota/discussions/17261) in this forum.

**Reply by NsinghP** [reply]
> You didn't get the point about `\"` Rule was at the right place in USER_RULE1. Why moving it to USER_BACKLOG and using a wrong syntax?

Ok, so for my understanding, I can declare as many rules as I want by using "USER_RULE1" " USER_RULE2" "USER_RUL..... lines to hardcode the rules into binary and then I have to use "USER_BACkLOG" command to execute these rules by calling semicolon separated "Rule1 1;Rule2 1;Rul.... along with semicolon separated commands that I want executed?

**Reply by sfromis** [reply]
You can have many rules in *each* of the 3 rule sets, up to `USER_RULE3`, and what you do with `Rule` commands in `USER_BACKLOG` is not to execute the rule sets, but to enable them, without that they are only stored but not active.

In general, I'd not recommend creating rules at build time,but instead via commands after first boot. One reason for this is that any changes has no effect when doing a normal upgrade to a new build, only if you also do a full reset.

**Reply by NsinghP** [reply]
> You can have many rules in _each_ of the 3 rule sets, up to `USER_RULE3`, and what you do with `Rule` commands in `USER_BACKLOG` is not to execute the rule sets, but to enable them, without that they are only stored but not active.
> 
> In general, I'd not recommend creating rules at build time,but instead via commands after first boot. One reason for this is that any changes has no effect when doing a normal upgrade to a new build, only if you also do a full reset.

Thanks for explaining, Yes, I noticed that I had to reset my device before the changes I made to binary come into effect. In this case burning the rules into binary is making it easy for me to flash a total of 7 devices I want to use for automating four 2-ton air conditioners and three electric water heaters. This device is very tiny and easily fits into the junction box of these appliances.

**Reply by sfromis** [reply]
Only a really small number like 7 devices means that you have used **way** more time messing with the builds, than just a couple of minutes to update all the devices using commands.

**Reply by NsinghP** [reply]
> Only a really small number like 7 devices means that you have used **way** more time messing with the builds, than just a couple of minutes to update all the devices using commands.

Haha, true, but now I am assured that if these devices reset due to power fluctuations I will be able to bring them back online very fast and the fact that after few months my memory of these strings will start fading helps in justifying the time spent. Plus I don't regret the time i spent discovering Tasmota secrets that are hiding in plain sight, as a mater of fact I love modding stuff in my free time. Actually I find playing with Tasmota to be a very stress relieving hobby.

EDIT: PS. I have ordered Sonoff  ZigBee Bridge Pro. I am thinking of converting all my Tasmota Devices from wifi to ZigBee. I think it can be done so I compiled my last build (for this Tinxy device) with ZigBee option! Grand Plan is to flash Sonoff Bridge and interface it with Home Assitant server and Tasmota Flashed Devices.

**Reply by NsinghP** [reply]
> > Question: Do I even need to integrate these into custom bin file or Should I make these changes after flashing the device through serial console?
> 
> You don't need. Instructions there are intended as commands to be entered in the console after flashing. There could be advantage to burn them into the binary but I suggest you first get experience with the device before going that way
> 
@barbudor In my opinion best way to flash this Tinxy 16Amp device is to use the custom binary file method. Let me explain my reasoning: If this device is configured using console commands then  in Post deployment scenario  (where it will be powered through Phase and Neutral port) when this device is reset, it will become functionally bricked until you retrieve the device and power it up by soldering wires to 3v3 and Gnd headers on the PCB. This will happen because  the device goes into 30 second reset loop after a reset and does not give enough time to enter Wifi credentials. (Except when Reset 4 or 5 or 6 command is used, but if it gets reset due to multiple (Six) power outages in brief period or due to other random fault, then it will be difficult to connect it back to Wifi). If you agree I will change the marked answer to custom binary method instead of console command method. I am assuming other Tinxy devices have same mechanism and all of these devices will become inoperable with off-the -shelf "Tasmota" firmware when they encounter their first "Reset" for whatever reason, during their lifetime.

**Reply by barbudor** [reply]
It's your choice
Building 1 binary per device is ESPhome philosophy. If you have 30 devices that's 30 binaries to rebuild on each release. 

Tasmota is IMHO more flexible as a single binary can fit multiple devices and configuration can easily be tweaked during the life of the device without the need to recompile a specific binary

I do burn some settings in my binaries but mostly only the essentials
- wifi credentials
- MQTT settings
- an auto-reconfigure rule based on WebGetConfig

The latest point is how I handle possible configuration resets. Should a device configuration be reset (willingly or not) it will automatically reconfigure itself from the last-known-good saved configuration file from a local web server.

As I said, Tasmota is flexible and you are free to choose how you want to use it...

**Reply by NsinghP** [reply]
> 
> As I said, Tasmota is flexible and you are free to choose how you want to use it...

Then, In the case of this Tinxy16Amp device, and possibly other devices by same manufacturer, it will be safe to say that "do not use" stock, off-the-shelf, "Tasmota' Firmware with configuration added through console after flashing, unless you are comfortable with occasional soldering when the the device resets. Recommended method of flashing this device is to use a customized "Tasmota" firmware binary which should either have hard coded "Wifi Credentials' or Hard coded custom strings: 

Either:

```
#define MODULE            USER_MODULE
#define USER_TEMPLATE "{\"NAME\":\"Tinxy16A-1\",\"GPIO\":[32,0,0,0,160,224,0,0,288,0,416,0,0,0],\"FLAG\":0,\"BASE\":18}"
#define STA_SSID1     "Enter Your Wifi router SSID"
#define STA_PASS1    "Enter Your router Wifi Password or Passphrase"
```
or
```
#define MODULE            USER_MODULE
#define USER_TEMPLATE "{\"NAME\":\"Tinxy16A-1\",\"GPIO\":[32,0,0,0,160,224,0,0,288,0,416,0,0,0],\"FLAG\":0,\"BASE\":18}"
#define LIGHT_MODE false 
#define LIGHT_CHANNEL_MODE false
#define USER_RULE1 "on system#boot do pwm1 512 endon"
#define USER_BACKLOG "Rule1 1;Pwmfrequency 1000;Pwmrange 1023 "

```
and
It may also be advisable to disable/modify official OTA Url to prevent inadvertent upgrade of custom binary to stock binary by adding following line to either one of the above custom string sets:

`#define OTA_URL "http://Disabled-please-consult-Network-Admin-/" PROJECT "-"  CODE_IMAGE_STR ".bin.gz"`

EDIT: @barbudor @sfromis :Please point out and correct any errors in my understanding of the issue and the code.

**Reply by barbudor** [reply]
> it will be safe to say that "do not use" stock, off-the-shelf, "Tasmota' Firmware with configuration added through console after flashing, unless you are comfortable with occasional soldering when the the device resets

Totally wrong.
When stock tasmota resets, it will re-open the Soft AP so you can reconnect to reconfigure. No soldering required.
You need to open the device only once.

There is a huge gap between making a custom build with just basic Wifi/MQTT settings that you can use with all of your devices (1 single binary rule them all) and making a dedicated binary for each device. 
If you save each device configuration, it's easy to reconfigure.
It's also easy to have an automatic reconfiguration like I did

But, that's YOUR choice. Don't imply it should be everyone's path.

**Reply by NsinghP** [reply]
> > it will be safe to say that "do not use" stock, off-the-shelf, "Tasmota' Firmware with configuration added through console after flashing, unless you are comfortable with occasional soldering when the the device resets
> 
> Totally wrong. When stock tasmota resets, it will re-open the Soft AP so you can reconnect to reconfigure. No soldering required. You need to open the device only once.
> 
You are right, in my case, Tasmota, did re-open the soft AP, but everytime I tried connecting to the soft AP and tried to enter the SSID and Passphrase the device timed out and reset itself. (In my case it is taking longer time to connect to Soft-AP as I have it protected by "#define WEB_PASSWORD" so technically in absence of "#define WEB_PASSWORD" if one types very fast it will be possible to connect the device to  Wifi Network.)
>
> There is a huge gap between making a custom build with just basic Wifi/MQTT settings that you can use with all of your devices (1 single binary rule them all) and making a dedicated binary for each device. If you save each device configuration, it's easy to reconfigure. It's also easy to have an automatic reconfiguration like I did
> 
For my other set of devices I am using custom binary with as many burned in common settings as possible and backing-up individual configuration files of each device to save unique settings local to each device. Set of these 16amp devices will have one common custom binary.
>
> But, that's YOUR choice. Don't imply it should be everyone's path.

Agreed, it was not my intent to imply everyone should use the settings that worked for me, I was just documenting my understanding of what worked for me, and thankyou for your time.

**Reply by sfromis** [reply]
My clear preference would be to actually understand the causes of that quick reset you are talking about, and deal with that issue, aiming for it to be workable without *needing* special builds with full config baked in.

**Reply by NsinghP** [reply]
> My clear preference would be to actually understand the causes of that quick reset you are talking about, and deal with that issue, aiming for it to be workable without _needing_ special builds with full config baked in.

Yes, agreed, and I think the issue mentioned in this post (https://github.com/arendst/Tasmota/discussions/17261) was causing the problem. EDIT: I don't fully understand but it has something to do with PWM signal at boot, and creating a continuous PWM signal on GPIO14 on initial boot solves the problem. I think that can only be done by burning PWM settings in the binary to make permanent changes so that after reset the PWM signal does not vanish.

**Reply by NsinghP** [reply]
> > My clear preference would be to actually understand the causes of that quick reset you are talking about, and deal with that issue, aiming for it to be workable without _needing_ special builds with full config baked in.
> 
> Yes, agreed, and I think the issue mentioned in this post (#17261) was causing the problem. EDIT: I don't fully understand but it has something to do with PWM signal at boot, and creating a continuous PWM signal on GPIO14 on initial boot solves the problem. I think that can only be done by burning PWM settings in the binary to make permanent changes so that after reset the PWM signal does not vanish.

I was able to recreate the reset problem by flashing stock tasmota on brand new device, entering PWM settings via the console and then unplugging the device from serial 3v3 and GND wires and  initiating a configuration Reset after powering the device using Mains Phase and Neutral . As expected the device entered a reset loop. As soon as I unplugged the device from mains and powered it using 3v3 and GND header the problem went away.

I then re-flashed the same device with my custom binary, again connected it to Mains Phase and neutral wire, issued a reset command from console  and the problem did not reoccur.

**Reply by NsinghP** [reply]
> You didn't get the point about `\"` Rule was at the right place in USER_RULE1. Why moving it to USER_BACKLOG and using a wrong syntax?

Replacing "on system#boot do pwm1 512 endon" to "on system#Init do pwm1 512 endon" Finally Solved the problem for me. However, as recommended, instead of complicated custom binary file method, eaiser way is to enter the changes using console commands, though one has to type reasonably fast when enterinig wifi SSID and Password.

**Reply by sfromis** [reply]
Copy/paste recommended, very fast for typing..... Can be useful to keep needed commands in a file.

**Reply by NsinghP** [reply]
> > You didn't get the point about `\"` Rule was at the right place in USER_RULE1. Why moving it to USER_BACKLOG and using a wrong syntax?
> 
> Replacing "on system#boot do pwm1 512 endon" to "on system#Init do pwm1 512 endon" Finally Solved the problem for me. However, as recommended, instead of complicated custom binary file method, eaiser way is to enter the changes using console commands, though one has to type reasonably fast when enterinig wifi SSID and Password.

Update: above discussed commands only solve the problem partially! If GPIO04 of the device is used as Switch 1 and mains live is fed to switch port of device the reset problem returns and device reboots every 10 seconds. Only solution is to not use Switch Port with mains Live input, instead use button 1 on GPIO0 to physically toggle on/off ( Please see https://github.com/arendst/Tasmota/discussions/18458 )
